### ISSUE TITLE
Ignore initial_node_count setting for Node Pool

### ISSUE THREAD

**Issue by mltsy** [terraform-providers/terraform-provider-google#6901]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

When the initial_node_count setting changes from a set value in the configuration file, this provider plans to destroy and recreate the entire node pool.  This is an issue because if you manually resize the node pool from the Cloud Console UI, that change updates the initial_node_count, so the next time you apply the Terraform config, it tries to recreate the whole node pool.

While it is possible to avoid this by omitting initial_node_count from your config, there is a use case for that setting (obviously, or it wouldn't exist) - in that it spins up the cluster faster and more reliably if you set it appropriately to begin with.

I have two alternative proposals for fixing this:
1. Just ignore initial_node_count after creating the node pool.  It really doesn't matter much after that point (although technically, it does affect how many nodes are initially provisioned when a new zone is added to the pool)
2. Rather than re-creating the entire cluster, resize the cluster to match the configured initial_node_count.

For more info on reproducing the issue, see bug #6889

Note: In my opinion this is actually a bug in the GKE API, but I'm not sure where to put that.

### New or Affected Resource(s)

* google_container_node_pool

### Potential Terraform Configuration

No new settings - see #6889 for an example of an affected config.

### References

* #6896 (a PR to warn about the current behavior, but it would be great to fix it instead)
* #6889 (original bug ticket)

**Comment by ctwilleager** [comment]
This is actually more of a significant problem than a simple enhancement. For my deployment mode, I am essentially forced to use the "initial_node_count" option because otherwise the provider spins up empty nodepools with no workers, despite Autoscaling being defined with a minimum node count of 1. If someone else comes in and changes the value via the UI or some other means, and I attempt to rerun my Terraform workflow against that project, Terraform insists on destroying the nodepool entirely. This poses a threat to production systems in that Terraform will simply destroy the entire nodepool, rather needlessly in my opinion, and render the entire production system offline until a the pool has been replaced.

In this case, I feel strongly that @mltsy suggestion #1 is the most appropriate. Terraform should not need to replace the entire node pool over a variance in initial_node_count and it should be ignored after the pool has been created.

**Comment by mltsy** [comment]
Yeah, I agree that this is more a bug than an enhancement, but after filing the bug (#6889) @edwardmedia suggested that since I was able to work around the bug, fixing it should be filed as an "enhancement" ticket.  Sounds like you're not able to work around it, so maybe the bug should be reopened - I don't know.  I am curious what your "deployment mode" is though, where it ends up spinning up 0 nodes for some reason.  I was able to remove initial_node_count and it seemed to default to 3 on GKE.

All that is beside the point, however.  Regardless of whether one can work around it, it should be fixed, I think.

**Comment by ctwilleager** [comment]
@mltsy I'm deleting the default node pool and then using Terraform's "google_container_node_pool" resource to deploy the actual node pools configured to our requirements. If I don't specify the "initial_node_count" value in the "gke_container_node_pool" resource, then Terraform provisions the node pools as empty, even if Autoscaling configuration has been defined with a minimum of 1 worker. 

However, I was able to work around the problem using a "lifecycle" block that ignores changes to the "initial_node_count". I was able to run Terraform to deploy the cluster with an "initial_node_count" of 1. I then changed the node count in the UI and reran Terraform plan. Terraform did not insist on replacing the nodepool and found no changes.

``

# Create Kubernetes standard node pool.
resource "google_container_node_pool" "gke-nodepool-standard-1" {
  name               = "gke-nodepool-standard-1"
  location           = var.region
  cluster            = google_container_cluster.gke-cluster.name
  max_pods_per_node  = var.gke_max_pods_per_node
  version            = var.gke_version
  initial_node_count = 1

  # Autoscaling config.
  autoscaling {
    min_node_count = var.gke_min_nodes_per_zone
    max_node_count = var.gke_max_nodes_per_zone
  }

  lifecycle {
    ignore_changes = [
      initial_node_count
    ]
  }

  # Management Config
  management {
    auto_repair  = true
    auto_upgrade = false
  }

  node_config {
    service_account = "gke-node-svc-acct@${var.project_id}.iam.gserviceaccount.com"
    preemptible     = false
    machine_type    = var.gke_node_machine_type
    disk_type       = "pd-standard"
    disk_size_gb    = var.gke_node_disk_size

    # Needed for correctly functioning cluster, see
    # https://www.terraform.io/docs/providers/google/r/container_cluster.html#oauth_scopes
    oauth_scopes = [
      "https://www.googleapis.com/auth/cloud-platform",
      "https://www.googleapis.com/auth/userinfo.email"
    ]

    metadata = {
      disable-legacy-endpoints = "true"
    }

    labels = {
      "cloud.google.com/gke-preemptible" = false
    }

    tags = ["kubernetes"]
  }
}

**Comment by mltsy** [comment]
Oooh, I didn't know that was an option!  I haven't used a lifecycle block before... well that's interesting.  I mean I still think that would be a good default behavior, but thank you for a great workaround!

**Comment by danawillow** [comment]
Note to the team: probably setSize is the way to go on this, but whoever takes it should investigate a bit how the API behaves.

**Comment by sureshyadav76** [comment]
Terraform deploy the GKE module, but instance group and instance template creation 3–3 are issues. please show this problem

**Comment by joe-a-dbt** [comment]
Related Google issue around initial node count when using autoscaling and standing up a new node pool - https://issuetracker.google.com/issues/467033281
